src/libsyncengine/update_detection/file_system_observer/remotefilesystemobserverworker.cpp

```diff
@@ -474,225 +478,247 @@ ExitCode RemoteFileSystemObserverWorker::sendLongPoll(bool &changes) {
-                LOGW_SYNCPAL_DEBUG(_logger, L"Unknown operation received on file: " << SyncName2WStr(name).c_str() << L" ("
-                                                                                    << Utility::s2ws(id).c_str() << L")");
-            }
+    Poco::JSON::Object::Ptr capabilitiesObj = actionObj->getObject(capabilitiesKey);
+    if (capabilitiesObj) {
+        if (!JsonParserUtility::extractValue(capabilitiesObj, canWriteKey, actionInfo.canWrite)) {
+            return ExitCodeBackError;
         }
     }
 
     return ExitCodeOk;
 }
 
-ExitCode RemoteFileSystemObserverWorker::accessRightsAdded(const std::string &action, const NodeId &nodeId, bool &rightsAdded,
-                                                           SyncTime &createdAt, SyncTime &modtime) {
-    if (action == accessRightInsert || action == accessRightUpdate || action == accessRightUserInsert ||
-        action == accessRightUserUpdate || action == accessRightTeamInsert || action == accessRightTeamUpdate ||
-        action == accessRightMainUsersInsert || action == accessRightMainUsersUpdate) {
-        bool hasRights = false;
-        ExitCode exitCode = hasAccessRights(nodeId, hasRights, createdAt, modtime);
-        rightsAdded = hasRights;
+ExitCode RemoteFileSystemObserverWorker::processAction(const SyncName &usedName, const ActionInfo &actionInfo,
+                                                       std::set<NodeId, std::equal_to<>> &movedItems) {
+    SnapshotItem item(actionInfo.nodeId, actionInfo.parentNodeId, usedName, actionInfo.createdAt, actionInfo.modtime,
+                      actionInfo.type, actionInfo.size, actionInfo.canWrite);
+
+    const auto hasRights = [=](const NodeId &nodeId, bool &rightsOk
+                               , SyncTime &createdAt, SyncTime &modtime, int64_t &size) -> ExitCode {
+        ExitCode exitCode = ExitCodeOk;
+        if (exitCode = getFileInfo(nodeId, rightsOk, createdAt, modtime, size); exitCode != ExitCodeOk) {
+            LOGW_SYNCPAL_WARN(_logger, L"Error while determining access rights on item: "
+                                           << SyncName2WStr(actionInfo.name).c_str() << L" ("
+                                           << Utility::s2ws(actionInfo.nodeId).c_str() << L")");
+            invalidateSnapshot();
+        }
         return exitCode;
-    }
+    };
+
+    // Process action
+    switch (actionInfo.actionCode) {
+        // Item added
+        case ActionCode::actionCodeAccessRightInsert:
+        case ActionCode::actionCodeAccessRightUpdate:
+        case ActionCode::actionCodeAccessRightUserInsert:
+        case ActionCode::actionCodeAccessRightUserUpdate:
+        case ActionCode::actionCodeAccessRightTeamInsert:
+        case ActionCode::actionCodeAccessRightTeamUpdate:
+        case ActionCode::actionCodeAccessRightMainUsersInsert:
+        case ActionCode::actionCodeAccessRightMainUsersUpdate: {
+            bool rightsOk = false;
+            SyncTime createdAt = 0;
```

Comment:
These 3 local variables are not used anywhere. I would recommend to isolate them in the body of the lambda function.

Reply:
Done 👍